### DISCUSSION TITLE
Parsing to return action-setfield

### DISCUSSION THREAD

**Discussion by seneschal2** [TiddlyWiki/TiddlyWiki5#5783] [Q&A]
Hi Folks,

Parse rule is working (parse tiddler/haystack for needles) and completing with:
                return [{
                        type: "action-setfield",
                        tag: "$action-setfield",
                        attributes: {
                                field   : {type: "string", value: 'needles'},
                                value   : {type: "string", value: 'needlesList'}
                        },
                        children: parseResults.tree
                        }];

Not sure this rusty newb can create a custom widget, but expecting action-setfield to work. (new needles field with string value of "needlesList")  (no joy) 

The widget should have no impact on the tiddler other than parsing for tiddler text and creating the "needles" field with list of the "needles".

Looking for recommendations.  Thank you!  This following-up on [Trying to keep my unique parse tree](https://groups.google.com/g/tiddlywikidev/c/hTcrOaQ7b7I/m/fv-jCAedAQAJ)

**Comment by joshuafontany** [comment]
What type of object is needlesList? If it is not a string, then the
set-action will fail.
```
value : {type: "string", value: needlesList.toString()}
```

That might work?

See the $action-setfield widget code:
https://tiddlywiki.com/#%24%3A%2Fcore%2Fmodules%2Fwidgets%2Faction-setfield.js
```

/*
Invoke the action associated with this widget
*/
SetFieldWidget.prototype.invokeAction = function(triggeringWidget,event) {
	var self = this,
		options = {};
	options.suppressTimestamp = !this.actionTimestamp;
	if((typeof this.actionField == "string") || (typeof this.actionIndex
== "string")  || (typeof this.actionValue == "string")) {
		this.wiki.setText(this.actionTiddler,this.actionField,this.actionIndex,this.actionValue,options);
	}
	$tw.utils.each(this.attributes,function(attribute,name) {
		if(name.charAt(0) !== "$") {
			self.wiki.setText(self.actionTiddler,name,undefined,attribute,options);
		}
	});
	return true; // Action was invoked
};
```

Best,
Joshua Fontany

**Reply by seneschal2** [reply]
Hi Joshua,
needleslist is a string. action-setfield appears to do nothing in this use case

**Comment by pmario** [comment]
actions are only activated by events. So it's not enough to see what the parser returns. .. The context is important too.

**Comment by seneschal2** [comment]
"actions are only activated by events" -- thanks Mario.  How would I set a field value (create if necessary) based on the parser (wikiparser/rule)?  Perhaps I return a custom fieldsWidget?  What might that look like?

Thanks for your patience.

**Comment by saqimtiaz** [comment]
@seneschal2 if I understand correctly, you have a custom parse rule that parses your tiddlers and generates a parse tree. You then have a macro that can render this parse tree. 

Normally, the parse tree indicates widgets that should be user to render the content. Parsers cannot and should not set field values. The best approach would be to create a custom widget that does exactly what your macro does. That is, accepts your parse tree as a parameter and renders it. Then in your parse tree you return that widget with the parse tree as its parameter.

A shortcut here could be to return in your parser a macrocall widget which invokes your macro to render the content. The challenge would probably be in properly passing the parseTree as a string parameter.

See https://tiddlywiki.com/#MacroCallWidget

**Reply by seneschal2** [reply]
Hi Saq,

Thanks for your interest.  The wikified tiddler would still be displayed as expected.  The difference is the text would be parsed (syntax errors highlighted) and a successfully parsed, stringified parse tree would be value of a tiddler field.  The first time successfully parsed, the field would be created/set.   Future successful edits would update the field value.
I hope I am more profitable with this description.

I'm looking into the MacroCallWidget for use (or not) with a custom widget to set the tiddler field.  

Still the newb.
